scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -515,67 +536,197 @@ def dijkstra(csgraph, directed=True, indices=None,
-                           dist_matrix, predecessor_matrix, limitf)
+        if min_only:
+            _dijkstra_directed_multi(indices,
+                                     csr_data, csgraph.indices,
+                                     csgraph.indptr,
+                                     dist_matrix, predecessor_matrix,
+                                     source_matrix, limitf)
+        else:
+            _dijkstra_directed(indices,
+                               csr_data, csgraph.indices, csgraph.indptr,
+                               dist_matrix, predecessor_matrix, limitf)
     else:
         csgraphT = csgraph.T.tocsr()
         if unweighted:
             csrT_data = csr_data
         else:
             csrT_data = csgraphT.data
-        _dijkstra_undirected(indices,
-                             csr_data, csgraph.indices, csgraph.indptr,
-                             csrT_data, csgraphT.indices, csgraphT.indptr,
-                             dist_matrix, predecessor_matrix, limitf)
+        if min_only:
+            _dijkstra_undirected_multi(indices,
+                                       csr_data, csgraph.indices,
+                                       csgraph.indptr,
+                                       csrT_data, csgraphT.indices,
+                                       csgraphT.indptr,
+                                       dist_matrix, predecessor_matrix,
+                                       source_matrix, limitf)
+        else:
+            _dijkstra_undirected(indices,
+                                 csr_data, csgraph.indices, csgraph.indptr,
+                                 csrT_data, csgraphT.indices, csgraphT.indptr,
+                                 dist_matrix, predecessor_matrix, limitf)
 
     if return_predecessors:
-        return (dist_matrix.reshape(return_shape),
-                predecessor_matrix.reshape(return_shape))
+        if min_only:
+            return (dist_matrix.reshape(return_shape),
+                    predecessor_matrix.reshape(return_shape),
+                    source_matrix.reshape(return_shape))
+        else:
+            return (dist_matrix.reshape(return_shape),
+                    predecessor_matrix.reshape(return_shape))
     else:
         return dist_matrix.reshape(return_shape)
 
 
-cdef _dijkstra_directed(
-            np.ndarray[ITYPE_t, ndim=1, mode='c'] source_indices,
-            np.ndarray[DTYPE_t, ndim=1, mode='c'] csr_weights,
-            np.ndarray[ITYPE_t, ndim=1, mode='c'] csr_indices,
-            np.ndarray[ITYPE_t, ndim=1, mode='c'] csr_indptr,
-            np.ndarray[DTYPE_t, ndim=2, mode='c'] dist_matrix,
-            np.ndarray[ITYPE_t, ndim=2, mode='c'] pred,
-            DTYPE_t limit):
-    cdef unsigned int Nind = dist_matrix.shape[0]
-    cdef unsigned int N = dist_matrix.shape[1]
-    cdef unsigned int i, k, j_source, j_current
-    cdef ITYPE_t j
+cdef _dijkstra_setup_heap_multi(FibonacciHeap *heap,
```

Comment:
There may be some benefit to decorating with `@cython.boundscheck(False)` based on the cython html report, but you'll know better than I if this is safe. I guess this function should only get called from C, so maybe it would be safe?

Reply:
I agree this is only called from C and so should be safe.  Added the bounds decorator.